### ISSUE TITLE
[basic.fundamental] Alignment and value representation of `nullptr_t`

### ISSUE THREAD

**Issue by frederick-vs-ja** [cplusplus/CWG#644]
Full name of submitter (unless configured in github; will be published with the issue): Jiang An

Reference (section label): [basic.fundamental]

Link to reflector thread (if any):

Issue description:

`nullptr_t` has several properties in C23 that is either not specified or different in C++.

[WG14 N3220](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf) 7.21.2 p3 reads:
> 3. The size and alignment of `nullptr_t` is the same as for a pointer to character type. An object representation of the value `nullptr` is the same as the object representation of a null pointer value of type `void*`. An lvalue conversion of an object of type `nullptr_t` with such an object representation has the value `nullptr`; if the object representation is different, the behavior is undefined.<sup>303)</sup>

In C++ it's only required that `sizeof(std::nullptr_t) == sizeof(void*)`. Do we want to specify that `alignof(std::nullptr_t) == alignof(void*)`? (The superficial wording difference from C doesn't matter due to [[basic.compound] p7](https://eel.is/c++draft/basic.compound#7).)

The UB specified in that paragraph isn't present in C++. In C++, conversions from `nullptr_t` effectively touch no bit, so the value representation of a `nullptr_t` object, which is currently unspecified in C++, doesn't matter in most uses. However, the decision on value representation of `nullptr_t` can lead to different results via `bit_cast`, and there're implementation divergence (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117727).

Perhaps it's better to specifiy that the value representation (and the valid set of value bits) of `nullptr_t` is implementation-defined.

Suggested resolution:

Modify [[basic.fundamental] p16](https://eel.is/c++draft/basic.fundamental#16) as indicated:
> [...] `sizeof(std​::​nullptr_t)` shall be equal to `sizeof(void*)` <ins>**and `alignof(std::nullptr_t)` shall be equal to `alignof(void*)`**</ins>. <ins>**The value representation of `std::nullptr_t` and the non-empty set of value representations of `std::nullptr_t` that correspond to values are implementation-defined. [_Note ?_: An implementation can decide that the value representation of `std::nullptr_t` contains no bits. — _end note_]**</ins>

**Comment by jensmaurer** [comment]
I agree about the alignment question, I think we should specify this to be the same as `void*`, too.

Since C++ opted not to ever read the bits of the value representation of `std::nullptr_t` for the lvalue-to-rvalue conversion, I don't think we need to talk about its value representation, and there is no need to require an implementation to document its choice ("implementation-defined"). The results from `bit_cast` are highly implementation-dependent anyway, so I'm not seeing a problem with the implementation divergence.

**Comment by jensmaurer** [comment]
CWG2966

**Comment by eisenwave** [comment]
The proposed resolution appears incompatible with MSVC. https://godbolt.org/z/EWrnohWcE

GCC and Clang use the same alignment as `void*`, whereas MSVC uses an alignment of `1`. The proposed resolution is an ABI break, although only for absolutely unhinged code that uses `std::nullptr_t` at ABI boundaries.

**Comment by jensmaurer** [comment]
Well, if there is any breakage of existing implementations, it is already caused by C in the quoted section: "The size and alignment of nullptr_t is the same as for a pointer to character type."

**Comment by eisenwave** [comment]
It's arguably possible for an implementation to have a different alignment of `nullptr_t` in C and C++, although that makes the type pretty much unusable at cross-language ABI boundaries unless it goes into a register, which it very often does.

In any case, I would be more comfortable with the change if it was standardizing existing practice instead of breaking one of the implementations. MSVC doesn't implement C23 `nullptr` yet, and it's unclear when or if that happens. Maybe what comes out of it is a defect report for C, and C2y is going to relax the requirement. Then we'd look pretty silly by forcing an unnecessary MSVC ABI break.

**Comment by frederick-vs-ja** [comment]
MSVC seemingly treats `nullptr_t` as something aligned as `void*` when calculating class layout currently. [Godbolt link](https://godbolt.org/z/qj6TPWf3v). So I guess ABI breaking cases would be rare (can still happen if one explicitly uses `alignas(nullptr_t)`).

Also, Clang sets `alignof(nullptr_t)` equal to `alignof(void*)` even when using MSVC ABI...
